### PR TITLE
Remove Bluebird

### PR DESCRIPTION
**Pull request by Munter**
These days we shouldn't really need Bluebird anymore.

- [x] Replace bluebird with native promises where possible
- [x] Use `util.promisify` to promisify in stead of bluebird
- [x] Replace bluebird usage for concurrency. Maybe `p-] limit`?
- [x] Replace bluebird usage for `.finally` support in https://github.com/assetgraph/assetgraph/blob/master/lib/util/determineFileType.js#L12
- [x] Delete bluebird from package.json

I renamed the remaining uses of bluebird from `Promise` to `Bluebird` to make it easier to spot the leftover usages.

The `.finally` usage in https://github.com/assetgraph/assetgraph/blob/master/lib/util/determineFileType.js#L12 is a homegrown concurrency limiter. Could that whole thing also be replaced with `p-limit` or another out of the box concurrency control for native promises?

### CODE REVIEW

File: lib/transforms/addDataVersionAttributeToHtmlElement.js [hunk 1]
```diff
@@ -1,20 +1,17 @@
...
       // Try to derive a version tag from git:
       try {
-        const [stdout] = await Promise.fromNode(
-          cb =>
-            childProcess.exec(
-              'git describe --long --tags --always --dirty',
-              cb
-            ),
-          { multiArgs: true }
+        const rawVersion = childProcess.execSync(
```
Comment (papandreou) [inline lib/transforms/addDataVersionAttributeToHtmlElement.js]: Ouch, could we avoid `execSync`? Maybe something like:

```diff
diff --git a/lib/transforms/addDataVersionAttributeToHtmlElement.js b/lib/transforms/addDataVersionAttributeToHtmlElement.js
index 3debad62..bf958461 100644
--- a/lib/transforms/addDataVersionAttributeToHtmlElement.js
+++ b/lib/transforms/addDataVersionAttributeToHtmlElement.js
@@ -1,16 +1,20 @@
 const childProcess = require('child_process');
+const { promisify } = require('util');
 
 module.exports = (queryObj, version) => {
   return async function addDataVersionAttributeToHtmlElement(assetGraph) {
     if (typeof version === 'undefined') {
       // Try to derive a version tag from git:
       try {
-        const rawVersion = childProcess.execSync(
-          'git describe --long --tags --always --dirty',
-          {
-            encoding: 'utf-8'
-          }
-        );
+        const rawVersion = await promisify(cb =>
+          childProcess.exec(
+            'git describe --long --tags --always --dirty',
+            {
+              encoding: 'utf-8'
+            },
+            cb
+          )
+        )();
         version = rawVersion.replace(/^[\s\n\r]+|[\s\r\n]+$/g, '');
       } catch (e) {}
     }
```

File: lib/transforms/addPrecacheServiceWorker.js [hunk 2]
```diff
@@ -214,7 +214,7 @@ module.exports = (
     } else {
       batches = entryPointHtmlAssets.map(asset => [asset]);
     }
-    await Promise.map(batches, generateAndAttachPrecacheServiceWorkerAsync, {
+    await Bluebird.map(batches, generateAndAttachPrecacheServiceWorkerAsync, {
```
Comment (papandreou) [inline lib/transforms/addPrecacheServiceWorker.js]: https://github.com/sindresorhus/p-map is a drop-in replacement.

File: lib/transforms/drawGraph.js [hunk 3]
```diff
@@ -139,19 +138,19 @@ module.exports = targetFileName => {
     ]);
     dotProcess.stdin.end(dotSrc);
     dotProcess.stderr.on('data', chunk => console.warn(`dot STDERR: ${chunk}`));
-    await Promise.fromNode(cb => {
+    await new Promise(resolve => {
       dotProcess.on('error', error => {
         if (error.code === 'ENOENT') {
-          cb(new Error('"dot" not found. Please install graphviz.'));
+          throw new Error('"dot" not found. Please install graphviz.');
```
Comment (papandreou) [inline lib/transforms/drawGraph.js]: These `throw`s will turn into unhandled exceptions. They should reject the promise instead.

File: lib/transforms/gzip.js [hunk 4]
```diff
@@ -9,25 +10,23 @@ module.exports = queryObj => {
...
-    await Promise.map(
+    await Bluebird.map(
       gzipAssets,
       async asset => {
         // http://webmasters.stackexchange.com/questions/31750/what-is-recommended-minimum-object-size-for-gzip-performance-benefits
         if (asset.type !== 'SourceMap' && asset.rawSrc.length > 860) {
-          const gzippedRawSrc = await Promise.fromNode(cb =>
-            compress.gzip(asset.rawSrc, cb)
-          );
+          const gzippedRawSrc = await compress.gzip(asset.rawSrc);
```
Comment (papandreou) [inline lib/transforms/gzip.js]: Hmm, does this work? It seems weird that `compress.gzip` would return a promise when only the main export of `node-zopfli` or `gzip` has been promisified?

We don't seem to have any tests covering this transform.

File: lib/transforms/loadAssets.js [hunk 5]
```diff
@@ -10,7 +11,7 @@ module.exports = function(...args) {
   const flattenedArguments = _.flatten(args);
   return async function loadAssets(assetGraph) {
     return _.flatten(
-      await Promise.map(
+      await Bluebird.map(
```
Comment (papandreou) [inline lib/transforms/loadAssets.js]: `p-map` should work fine here, too.

File: lib/transforms/gzip.js (hunk 4 above)
Comment (Munter) [reply lib/transforms/gzip.js]: Good catch. Fixed in aa133bbb

File: lib/transforms/drawGraph.js (hunk 3 above)
Comment (Munter) [reply lib/transforms/drawGraph.js]: Fixed in dcd95269

File: lib/transforms/addDataVersionAttributeToHtmlElement.js (hunk 1 above)
Comment (Munter) [reply lib/transforms/addDataVersionAttributeToHtmlElement.js]: Fixed in 43a1ae24

Why doesn't this work?

```js
const rawVersion = await promisify(childProcess.exec)(
  'git describe --long --tags --always --dirty',
  {
    encoding: 'utf-8'
  }
);
```

Comment (papandreou) [reply lib/transforms/addDataVersionAttributeToHtmlElement.js]: Hmm, yeah, that's more elegant than involving a callback. It does work if you destructure the fulfillment value:

```js
        const { stdout } = await promisify(childProcess.exec)(
          'git describe --long --tags --always --dirty',
          {
            encoding: 'utf-8'
          }
        );
        version = stdout.replace(/^[\s\n\r]+|[\s\r\n]+$/g, '');
```

File: lib/transforms/addPrecacheServiceWorker.js (hunk 2 above)
Comment (Munter) [reply lib/transforms/addPrecacheServiceWorker.js]: Done in 4fd02081

File: lib/transforms/loadAssets.js (hunk 5 above)
Comment (Munter) [reply lib/transforms/loadAssets.js]: Done in 4fd0208

File: lib/transforms/addDataVersionAttributeToHtmlElement.js (hunk 1 above)
Comment (Munter) [reply lib/transforms/addDataVersionAttributeToHtmlElement.js]: Ah! I'll make that change.

Comment (Munter) [reply lib/transforms/addDataVersionAttributeToHtmlElement.js]: Done in b810eab7

File: lib/transforms/gzip.js (hunk 4 above)
Comment (papandreou) [reply lib/transforms/gzip.js]: Great! I added a gzip test on master (https://github.com/assetgraph/assetgraph/commit/eb007153b08ec91149bd684ff0b0a2dbb307f559), and it also passes on this branch :muscle:

**Comment by papandreou** [comment]
@Munter, I tried to replace that hand-rolled concurrency limiter in `determineFileType` with `p-limit` + added a test in 2643cd2.

**Comment by Munter** [comment]
Yay! Great to slim things down a bit. Then we just need https://github.com/One-com/teepee/pull/8 merged so we can get bluebird completely out of the non-development

**Pull request merged**